hollow/src/main/java/com/netflix/hollow/core/memory/encoding/FixedLengthMultipleOccurrenceElementArray.java

```diff
@@ -0,0 +1,161 @@
+/*
+ *
+ *  Copyright 2018 Netflix, Inc.
+ *
+ *     Licensed under the Apache License, Version 2.0 (the "License");
+ *     you may not use this file except in compliance with the License.
+ *     You may obtain a copy of the License at
+ *
+ *         http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *     Unless required by applicable law or agreed to in writing, software
+ *     distributed under the License is distributed on an "AS IS" BASIS,
+ *     WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *     See the License for the specific language governing permissions and
+ *     limitations under the License.
+ *
+ */
+package com.netflix.hollow.core.memory.encoding;
+
+import com.netflix.hollow.core.memory.pool.ArraySegmentRecycler;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.LongStream;
+
+/**
+ * This class stores multiple instances of a fixed bit width element in a list of nodes. For
+ * example, it can store multiple 6 bit elements at different indices of the list of nodes.
+ *
+ * Under the hood, it uses a {@link FixedLengthElementArray} in order to implement compact storage
+ * that allows inserting multiple instances of the fixed length elements. It maintains at least
+ * enough space to hold up to the max number of occurrences of each element. This means that if
+ * there are 8 elements at one node index, it will contain enough space to store at least 8 elements
+ * at each index.
+ *
+ * Running out of space at any index triggers a relatively expensive resize operation, where we
+ * create storage of a multiple (currently 1.5x) of the previous storage and copy items over, but
+ * this can be alleviated by passing in a better guess for max elements per node.
+ * Note that this class is currently designed to be used for a relatively small number for
+ * bitsPerElement.
+ */
+public class FixedLengthMultipleOccurrenceElementArray {
+    private static final double RESIZE_MULTIPLE = 1.5;
+    private static final long NO_ELEMENT = 0L;
+
+    private final ArraySegmentRecycler memoryRecycler;
+    private final FixedLengthElementArray nodesWithOrdinalZero;
+    private final int bitsPerElement;
+    private final long elementMask;
+    private final long numNodes;
+
+    private volatile FixedLengthElementArray storage;
+    private volatile int maxElementsPerNode;
+
+    public FixedLengthMultipleOccurrenceElementArray(ArraySegmentRecycler memoryRecycler,
+            long numNodes, int bitsPerElement, int maxElementsPerNodeEstimate) {
+        nodesWithOrdinalZero = new FixedLengthElementArray(memoryRecycler, numNodes);
```

Comment:
Interesting... Something like storing the current size for each node in an int[]?

Reply:
Hm, we might be able make that work for our purposes because the number of bits needed to represent the size is the same number of bits needed to represent the long elements themselves, because in our case they are ordinals, and the size is bounded by the highest ordinal.
The only issue then is duplicates - we need to de-dupe in our storage to ensure that the size of elements stored per nodeIndex cannot grow beyond its upper bound (1 << numBitsPerElement - 1) - right now we don't handle duplicates and that condition does not hold.